source/GameInterface/Services/Armies/Patches/ArmyPatches.cs

```diff
@@ -16,45 +20,87 @@ namespace GameInterface.Services.Armies.Patches;
     [HarmonyPrefix]
-    static bool OnAddPartyInternalPrefix(ref Army __instance, ref MobileParty mobileParty)
+    static bool OnAddPartyInternalPrefix(ref Army __instance, MobileParty mobileParty)
     {
         if (AllowedThread.IsThisThreadAllowed()) return true;
         if (PolicyProvider.AllowOriginalCalls) return true;
 
         if (ModInformation.IsClient) return false;
+        
+
+        if (ContainerProvider.TryResolve<IObjectManager>(out var objectManager) == false)
+        {
+            Logger.Error("Unable to resolve {objectManager}", typeof(IObjectManager));
+            return false;
+        }
 
-        var message = new MobilePartyInArmyAdded(mobileParty.StringId, __instance.LeaderParty.StringId);
+        string armyId = __instance.GetStringId();
+        if (armyId == null)
+        {
+            Logger.Error("{army} was not properly registered", mobileParty.Army.Name);
+            return false;
+        }
+
+        var message = new MobilePartyInArmyAdded(mobileParty.StringId, armyId);
         MessageBroker.Instance.Publish(mobileParty, message);
 
         return true;
     }
 
+
     [HarmonyPatch(typeof(Army), "OnRemovePartyInternal")]
     [HarmonyPrefix]
-    static bool OnRemovePartyInternalPrefix(ref Army __instance, ref MobileParty mobileParty)
+    static bool OnRemovePartyInternalPrefix(ref Army __instance, MobileParty mobileParty)
     {
         if (AllowedThread.IsThisThreadAllowed()) return true;
         if (PolicyProvider.AllowOriginalCalls) return true;
 
         if (ModInformation.IsClient) return false;
+        
 
-        var message = new MobilePartyInArmyRemoved(mobileParty.StringId, __instance.LeaderParty.StringId);
+        if (ContainerProvider.TryResolve<IObjectManager>(out var objectManager) == false)
+        {
+            Logger.Error("Unable to resolve {objectManager}", typeof(IObjectManager));
+            return false;
+        }
+        
         
+        string armyId = __instance.GetStringId();
+        if (armyId == null)
+        {
+            Logger.Error("{army} was not properly registered", mobileParty.Army.Name);
+            return false;
+        }
+
+        var message = new MobilePartyInArmyRemoved(mobileParty.StringId, armyId);
+
         MessageBroker.Instance.Publish(mobileParty, message);
+        
 
         return true;
     }
 
+
     public static void AddMobilePartyInArmy(MobileParty mobileParty, Army army)
     {
         GameLoopRunner.RunOnMainThread(() =>
         {
             using (new AllowedThread())
             {
+                /*foreach (var mobileParty_local in army.Parties)
+                {
+                    // Apply this code if duplication found
+                    if (mobileParty_local.StringId == mobileParty.StringId)
+                    {
+                        return;
+                    }
+                }*/
                 ArmyExtensions.AddPartyInternal(mobileParty, army);
+
```

Comment:
Remove dead code and use army extension
```suggestion
                army.AddPartyInternal(mobileParty);

```